### PR TITLE
The request System. Or as it is commonly known the DMan 2.0

### PR DESCRIPTION
**Pull request by marchermans**
This is a discussion PR.

Ist main goal is to get Input on the design of the new System.
As you all can see it is heavily Interface and factory based.

The principle follows an Inversion of Control pattern that makes the IRequestManager the heart of the System. Outside of him no IRequests are known, just Tokens and These Tokens are passed around.

If you request a IRequest from the Manager you will get a defensive copy to make sure that you cannot tamper with its state directly.

RequestResolvers are the workers in this System. A resolver basicly tells the Handler what to do. And how to proceed.

Please feel free to ask me anything regarding the new System.

### CODE REVIEW

**Comment by marchermans** [comment]
Yes there are a bunch of Things that i Need to merge. This pr is just for discussion and should be treated as such.

File: src/api/java/com/minecolonies/coremod/colony/permissions/Action.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    SEND_MESSAGES(6),
+    EDIT_PERMISSIONS(7),
+    MANAGE_HUTS(8);
+
+    private final int flag;
+
+    /**
+     * Stores the action as byte.
+     * {@link #ACCESS_HUTS} has value 0000 0000
+     * {@link #SEND_MESSAGES} has value 0100 0000
```
Comment (marvin-bitterlich) [inline src/api/java/com/minecolonies/coremod/colony/permissions/Action.java]: showing only 8 of 32 bits makes it seem like a bug ^^

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+     * Index list used to read and write from NBT
+     */
+    static ArrayList<RequestState> indexList = new ArrayList<>();
+
+    static {
+        /**
+         * This should never be changed! It is used to read and write from NBT so it has to
+         * persist between mod versions.
+         */
+        for(RequestState state : RequestState.values())
```
Comment (marvin-bitterlich) [inline src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java]: optional braces?

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    @NotNull
+    IFactoryController getFactoryController();
+
+    /**
+     * Method to create a request for a given object
+     * @param requester The requester.
+     * @param object The Object that is being requested.
+     * @param <T> The type of request.
+     * @return The token representing the request.
+     * @throws IllegalArgumentException is thrown when this manager cannot produce a request for the given types.
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java]: Should we really throw illegalArgument exceptions or should we return something like Token.INVALID?

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * State for a request that has been completed.
+     */
+    COMPLETED,
+
+    /**
+     * State for a request when it has been forcefully fullfilled by a player
+     */
+    OVERRULED,
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java]: Nice, the states are great now.

Comment (Raycoms) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java]: The formatting of this file is off

File: src/main/java/com/minecolonies/coremod/colony/Colony.java [hunk 5]
```diff
@@ -1436,8 +1440,19 @@ public void addWayPoint(final BlockPos point, IBlockState block)
      * @return Map with ID (coordinates) as key, and buildings as value.
      */
     @NotNull
-    public Map<BlockPos, AbstractBuilding> getBuildings()
+    public ImmutableMap<BlockPos, IBuilding> getBuildings()
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/Colony.java:1652]: does it have a reason you return Immuteablemap?

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.minecolonies.coremod.colony.requestsystem.location;
+
+import net.minecraft.util.math.BlockPos;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Interface used to describe locations in the world.
+ */
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java]: Colonist are currently unable to switch dimensions, should take care with that, it causes a lot of problems with them atm.

File: src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java [hunk 7]
```diff
@@ -154,12 +150,7 @@ public static AbstractJob createFromNBT(final CitizenData citizen, @NotNull fina
      */
     public void readFromNBT(@NotNull final NBTTagCompound compound)
     {
-        final NBTTagList itemsNeededTag = compound.getTagList(TAG_ITEMS_NEEDED, Constants.NBT.TAG_COMPOUND);
-        for (int i = 0; i < itemsNeededTag.tagCount(); i++)
-        {
-            final NBTTagCompound itemCompound = itemsNeededTag.getCompoundTagAt(i);
-            itemsNeeded.add(new ItemStack(itemCompound));
-        }
+        //NOOP; Requests are stored on the building.
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java:163]: Nice 😍

File: src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java [hunk 8]
```diff
@@ -714,73 +714,6 @@ public void serializeViewNetworkData(@NotNull final ByteBuf buf, @NotNull final
     }
 
     /**
-     * Ranks within a colony.
-     */
-    public enum Rank
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java]: 👍

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/StandardFactoryController.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+import java.util.HashMap;
+
+/**
+ * Default implementation of a FactoryController
+ * Singleton.
+ */
+public class StandardFactoryController implements IFactoryController {
+
+    ////// --------------------------- NBTConstants --------------------------- \\\\\\
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/StandardFactoryController.java]: Cool 👓

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java [hunk 10]
```diff
@@ -1078,13 +1079,54 @@ else if (block instanceof BlockFlowerPot)
...
         }
 
-        if (block instanceof BlockChest && job instanceof JobBuilder)
-        {
-            final BlockPos buildingLocation = ((JobBuilder) job).getWorkOrder().getBuildingLocation();
-            final AbstractBuilding building = this.getOwnBuilding().getColony().getBuilding(buildingLocation);
-            building.addContainerPosition(pos);
+
+        try{
+            if (job instanceof JobBuilder) {
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: can we make this go away? ^^

File: src/main/resources/assets/minecolonies/lang/en_us.lang [hunk 11]
```diff
@@ -215,6 +217,3 @@ tile.minecolonies.blockHutWareHouse.name=Warehouse
 com.minecolonies.coremod.gui.workerHuts.buildingWareHouse=Warehouse
 tile.minecolonies.blockHutDeliveryman.name=Deliveryman
 com.minecolonies.coremod.gui.workerHuts.Deliveryman=Deliveryman
-com.minecolonies.coremod.gui.workerHuts.cancelBuild=Cancel Build
-com.minecolonies.coremod.gui.workerHuts.cancelUpgrade=Cancel Upgrade
```
Comment (marvin-bitterlich) [inline src/main/resources/assets/minecolonies/lang/en_us.lang]: why?

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java (hunk 10 above)
Comment (xavierh) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: Its in my pr already I think!

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java (hunk 4 above)
Comment (marvin-bitterlich) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java]: formatting is off everywhere, @OrionDevelopment will fix it prior to merge

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java (hunk 10 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java]: That should be away already I think. I rewrote all that builder stuff already.
Probably just has to merge the newest changes to his branch

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java (hunk 6 above)
Comment (marvin-bitterlich) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java]: still fine to record what dim we are in

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requestable/Tool.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
+package com.minecolonies.coremod.colony.requestsystem.requestable;
+
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Class used to represent tools inside the request system.
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requestable/Tool.java]: Nice, we really needed that

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java (hunk 6 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java]: Just noted we need to take care of that =D

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java (hunk 3 above)
Comment (marvin-bitterlich) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java]: exeption is better style then null value imo

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requestable/Tool.java (hunk 12 above)
Comment (marvin-bitterlich) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/requestable/Tool.java]: 🎉

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java (hunk 3 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java]: I wouldn't make it a null value. I would make it really a Token.Invalid with a concrete value.
We had a lot of exceptions in minecolonies which we removed almost completely because they did more bad than good.

File: src/main/resources/assets/minecolonies/lang/en_us.lang (hunk 11 above)
Comment (Raycoms) [reply src/main/resources/assets/minecolonies/lang/en_us.lang]: Probably because old prs are not merged yet.

File: src/main/java/com/minecolonies/coremod/colony/Colony.java (hunk 5 above)
Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/Colony.java:1652]: Why would it not Need to be an Immutable. Do you want something from outside your colony to mess up the map, ofcourse not. If it would not be an immutable, somebody might remove a Building and the request System does not get notified.

File: src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java (hunk 7 above)
Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java:163]: :P It was one of the Corner Stones. Request are state based, so hey are stored in something else.

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java (hunk 3 above)
Comment (marchermans) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java]: I would like to see which type you would use?

File: src/main/java/com/minecolonies/coremod/colony/Colony.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/Colony.java:1652]: There is also the option of "Unmodifiable map"

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java (hunk 3 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java]: Its an IToken but it has certain values so it equals IToken.INVALID.

Comment (marchermans) [reply src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java]: Itoken is an interface. Technically just hull around a different object. The standard uses a uuid. However i think this is better it is stated in the method header that this method throws the exception so everybody knows it does.

File: src/api/java/com/minecolonies/coremod/colony/IColony.java [hunk 13]
```diff
@@ -61,4 +66,27 @@
...
+    @Nullable
+    World getWorld();
+
+    /**
+     * Returns the buildings in the colony
+     *
+     * @return The buildings in the colony
+     */
+    @NotNull
+    ImmutableMap<BlockPos, IBuilding> getBuildings();
```
Comment (xavierh) [inline src/api/java/com/minecolonies/coremod/colony/IColony.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/api/java/com/minecolonies/coremod/colony/buildings/IBuilding.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
+package com.minecolonies.coremod.colony.buildings;
+
+import com.minecolonies.coremod.colony.IColony;
+import net.minecraft.util.math.BlockPos;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Created by marcf on 3/7/2017.
+ */
+public interface IBuilding {
```
Comment (xavierh) [inline src/api/java/com/minecolonies/coremod/colony/buildings/IBuilding.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/api/java/com/minecolonies/coremod/colony/permissions/Action.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
+package com.minecolonies.coremod.colony.permissions;
```
Comment (xavierh) [inline src/api/java/com/minecolonies/coremod/colony/permissions/Action.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/api/java/com/minecolonies/coremod/colony/permissions/IPermissions.java [hunk 16]
```diff
@@ -14,10 +14,10 @@
      * Returns whether the player has the permission for an action.
      *
      * @param player {@link EntityPlayer} player.
-     * @param action {@link com.minecolonies.coremod.colony.permissions.Permissions.Action} action.
+     * @param action {@link Action} action.
      * @return true if has permission, otherwise false.
      */
-    boolean hasPermission(EntityPlayer player, Permissions.Action action);
+    boolean hasPermission(EntityPlayer player, Action action);
```
Comment (xavierh) [inline src/api/java/com/minecolonies/coremod/colony/permissions/IPermissions.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/api/java/com/minecolonies/coremod/colony/permissions/Rank.java [hunk 17]
```diff
@@ -0,0 +1,33 @@
+package com.minecolonies.coremod.colony.permissions;
```
Comment (xavierh) [inline src/api/java/com/minecolonies/coremod/colony/permissions/Rank.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocatable.java [hunk 18]
```diff
@@ -0,0 +1,16 @@
+package com.minecolonies.coremod.colony.requestsystem.location;
```
Comment (xavierh) [inline src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocatable.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
+package com.minecolonies.coremod.colony.requestsystem.location;
```
Comment (xavierh) [inline src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/main/java/com/minecolonies/coremod/blocks/AbstractBlockHut.java [hunk 20]
```diff
@@ -202,7 +202,7 @@ public boolean onBlockActivated(
 
             if (building != null
                   && building.getColony() != null
-                  && building.getColony().getPermissions().hasPermission(playerIn, Permissions.Action.ACCESS_HUTS))
+                  && building.getColony().getPermissions().hasPermission(playerIn, Action.ACCESS_HUTS))
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/blocks/AbstractBlockHut.java:208]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java [hunk 21]
```diff
@@ -302,7 +303,7 @@ private void updateUsers()
     {
         users.clear();
         users.addAll(townHall.getColony().getPlayers().values());
-        users.sort(Comparator.comparing(Permissions.Player::getRank, Permissions.Rank::compareTo));
+        users.sort(Comparator.comparing(Permissions.Player::getRank, Rank::compareTo));
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java [hunk 22]
```diff
@@ -653,7 +654,7 @@ private void removePlayerClicked(final Button button)
         if (row >= 0 && row < users.size())
         {
             final Permissions.Player user = users.get(row);
-            if (user.getRank() != Permissions.Rank.OWNER)
+            if (user.getRank() != Rank.OWNER)
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java [hunk 23]
```diff
@@ -20,7 +23,7 @@
 /**
  * Extra data for Citizens.
  */
-public class CitizenData
+public class CitizenData implements ILocatable
```
Comment (xavierh) [inline src/main/java/com/minecolonies/coremod/colony/CitizenData.java]: I think this belong to a different pr, and could be merge well before dman 2.0 is ready

**Review by xavierh — CHANGES_REQUESTED** [review]
A lost of (good) stuff should be in a different PR and could be added well before the Request system is done
My old brain can not take in +4000 lines in one go.
some interface should be add earlier so that other devs can star using them.

**Comment by cltnschlosser** [comment]
I agree with @xavierh, there is way too much going on in this pull request. We need take a step back and get some info written down about what is going on here, and then create sub tasks with smaller goals. I'm sure there is a lot of valuable code here, but it's way to hard to tell. Let's get a plan together and start merging this piece by piece. It's in everyones best interest to not have long lived branches with tons of changes on them. I'd hate to see all this work get so separated from the mod that it can't be merged in.

This pull request also appears to be trying to separate the mod into an api (set of interfaces) and an implementation, but also implementing a request system. I think those should be done in separate pull requests.

Also, I think that if we are going to pull out an api, that instead of making it a separate package, we should make it a separate module, that our implementation depends on.

I would like to start moving some of our Enums and static maps to registries, which would be beneficial for api work, but it's hard for me to tell what has been done.

**Comment by marchermans** [comment]
So lets start from the beginning.
This Pull is an API, from scratch designed.
It will consist out of a Major rework and an even bigger refactoring.

# What does it contain when i am done:
## Module wise:
- A _Module_ for the API. It is needed to create an API Jar.
- A _Module_ for BlockOut. It is not needed, but creates a more cohesive package structure. Also would allow for better Distribution if we ever needed it again.
- A _Module_ for Structures. (Same as with the BlockOut Module)
## API wise:
- Interfaces and or Abstract classes for most of our Core classes, which include but are not limited too:
    - Colony and ColonyView -> IColony
    - AbstractBuilding And View -> IBuilding
    - AbstractJob -> IJob
    - etc.
- Utility Classes and the complete Lib-Package
- The Request System (at the heart of this PR, which is what it originally started as)
- A factory Pattern that should be used to create, serialize and deserialize new Objects that are managed by the API.
- A new ID System for Buildings (alltough rightnow i redirected all the calls to get ID() which returned a BlockPos and now Returns an IToken<UUID> to getLocation().getInDimensionLocation(), as getLocation() now Returns an ILocation, to be specific it returns a StaticLocation, hence the getInDimensionLocation Call). This new System is needed to identify Buildings properly inside the Request System, as from its standpoint the Building functions as the requester of a given request, even though it just manages the request for ist citizens.
## Refactoring wise:
- Due to the Need of the IColony Interface being in the API package, this caused a cascading downfall of classes either needing to be moved, renamed, replaced or extended somehow. A lot of refactoring also went into the readability of the Code. A lot of classes are entangled with others, without having an Interface in between to handle the interaction in proper OOP style. Refactoring this so i could use them properly is the biggest part of this PR in the first place.

# What is not in this PR:
- Registries for custom Objects. As of right now this is not part of this PR, because it is not directly needed for the Request System to function properly.
- Abstraction layers to make the API more stable. Yes these Need to be introduced.

# Why is this not done in small Pull Request.
- Well first of all, have you ever tried pulling IColony out the main Module and into an other one?. The amount of compile Errors it creates due to classes missing is enormous. Due to the lack of abstraction between the Interface and say AbstractBuilding (and all that that class depends upon) it is not possible to create a compiling Version of the Code. From the Moment I was done with the core API of the request System i have been working on getting it to compile, but i am still rewriting stuff to make the move of the IColony Interface work.
- Yes i Agree that their are some **small** changes that could have been (and maybe should be) in a different PR. However I am not a Person that commits after every file i have changed, nor should i be forced to when i make changes of this size. So i would like to see somebody cherrypick the files that he wants. If he can do it and make a PR he is more then welcome to do so.

TLDR: This is still not compiling because i moved IColony and it resulting in a Compiler outcry. And that is the reason why this is a big one push PR and not twenty different small ones.

**Comment by cltnschlosser** [comment]
Well cool, I'll start on the registries then. But even from looking at your bullet points I see 4-5 PRs

**Comment by marchermans** [comment]
I would like to ask you to wait with the registries, atleast untill i have this compiling. We can then discuss in how many PRS i Need to split this.
